### PR TITLE
Implemented PKCE Authorization Flow

### PR DESCRIPTION
**Pull request by ljubiknenad**
## What kind of change does this PR introduce?

This PR consist implementation of PKCE Authorization Flow for `getOAuthSignInURL` & `signInWithOTP` methods.

## What is the new behavior?

When initializing a client there is a new property called `flowType` which is of type FlowType enum that consist of two types:

- implicit
- pkce

example:
`let client = GoTrueClient(url: "", headers: [:], flowType: .pkce)`

The flowType property has default value of `.implicit`

### CODE REVIEW

File: Sources/GoTrue/GoTrueClient.swift [hunk 1]
```diff
@@ -23,17 +24,22 @@ public final class GoTrueClient {
       return try await env.sessionManager.session()
     }
   }
+    
+  public let flowType: FlowType!
```
Comment (grdsdev) [inline Sources/GoTrue/GoTrueClient.swift]: No need to be a unwrapped optional.

```suggestion
  public let flowType: FlowType
```

File: Sources/GoTrue/Types.swift [hunk 2]
```diff
@@ -526,3 +535,39 @@ public struct RecoverParams: Codable, Hashable, Sendable {
...
+    case implicit
+    case pkce
+}
+
+
+enum PKCE {
+    static func generateCodeVerifier() -> String {
+        var buffer = [UInt8](repeating: 0, count: 64)
+        _ = SecRandomCopyBytes(kSecRandomDefault, buffer.count, &buffer)
+        return Data(buffer).base64EncodedString()
```
Comment (grdsdev) [inline Sources/GoTrue/Types.swift]: Just confirming, should we use `pkceBase64EncodedString` here too?

File: Sources/GoTrue/Endpoints.swift [hunk 3]
```diff
@@ -12,6 +12,12 @@ extension Paths {
   struct Token {
     /// Path: `/token`
     let path: String
+      
+    func post(
+      grantType: GrantType,
+      body: PKCEParams) -> Request<GoTrue.Session> {
+        Request(method: "POST", url: path, query: makePKCEPostQuery(), body: body)
+      }
```
Comment (grdsdev) [inline Sources/GoTrue/Endpoints.swift]: I don't think we need another `post` method, just adding pkce options to `GrantType` and `PostRequest` should be enough, and then the GoTrueClient should reuse the already existing `post` method.

File: Sources/GoTrue/GoTrueClient.swift [hunk 4]
```diff
@@ -309,15 +335,41 @@ public final class GoTrueClient {
       throw error
     }
   }
+    
+    public func session(from url: URL, storeSession: Bool = true) async throws -> Session {
```
Comment (grdsdev) [inline Sources/GoTrue/GoTrueClient.swift:339]: Add docs and `discardableResult` to this method too, remove from the private one.

File: Sources/GoTrue/GoTrueError.swift [hunk 5]
```diff
@@ -5,6 +5,7 @@ public enum GoTrueError: LocalizedError, Sendable {
   case malformedJWT
   case sessionNotFound
   case api(APIError)
+  case codeNotFound
```
Comment (grdsdev) [inline Sources/GoTrue/GoTrueError.swift]:
```suggestion
  case pkceCodeNotFound
```

File: Sources/GoTrue/GoTrueError.swift [hunk 6]
```diff
@@ -28,6 +29,7 @@ public enum GoTrueError: LocalizedError, Sendable {
     case .malformedJWT: return "A malformed JWT received."
     case .sessionNotFound: return "Unable to get a valid session."
     case let .api(error): return error.errorDescription ?? error.message ?? error.msg
+    case .codeNotFound: return "Unable to get a valid code."
```
Comment (grdsdev) [inline Sources/GoTrue/GoTrueError.swift]:
```suggestion
    case .pkceCodeNotFound: return "A valid PKCE code was not found."
```

File: Sources/GoTrue/Types.swift [hunk 7]
```diff
@@ -384,19 +385,25 @@ public struct OTPParams: Codable, Hashable, Sendable {
   public var createUser: Bool
   public var data: [String: AnyJSON]?
   public var gotrueMetaSecurity: GoTrueMetaSecurity?
+  public var codeChallenge: String?
+  public var codeChallengeMethod: String?
```
Comment (grdsdev) [inline Sources/GoTrue/Types.swift]: Would be nice to have a `enum` representing the codeChallgendMethod, even we if we have single method.

```swift
public enum PKCECodeChallengeMethod: String {
    case sha256 = "S256"
}
```

**Review by grdsdev — CHANGES_REQUESTED** [review]

File: Sources/GoTrue/Types.swift (hunk 2 above)
Comment (ljubiknenad) [reply Sources/GoTrue/Types.swift]: Yup. As i saw the same is on Android

File: Sources/GoTrue/Endpoints.swift [hunk 8]
```diff
@@ -27,22 +27,32 @@ extension Paths {
       encoder.encode(redirectTo, forKey: "redirect_to")
       return encoder.items
     }
+    
+    private func makePKCEPostQuery() -> [(String, String?)] {
+      let encoder = URLQueryEncoder()
+      encoder.encode(GrantType.pkce, forKey: "grant_type")
+        
+      return encoder.items
+    }
```
Comment (grdsdev) [inline Sources/GoTrue/Endpoints.swift]: Remove as it isn't being used

**Review by grdsdev — CHANGES_REQUESTED** [review]

File: Sources/GoTrue/Types.swift [hunk 9]
```diff
@@ -526,3 +535,43 @@ public struct RecoverParams: Codable, Hashable, Sendable {
     case gotrueMetaSecurity = "gotrue_meta_security"
   }
 }
+
+public struct PKCEParams: Codable, Hashable, Sendable {
```
Comment (grdsdev) [inline Sources/GoTrue/Types.swift]: I'd rename this for consistency.

```suggestion
public struct PKCECredentials Codable, Hashable, Sendable {
```

**Review by grdsdev — CHANGES_REQUESTED** [review]

File: Sources/GoTrue/Types.swift (hunk 2 above)
Comment (grdsdev) [reply Sources/GoTrue/Types.swift]: Didn't get it, should we use `pkceBase64EncodedString`?

Comment (ljubiknenad) [reply Sources/GoTrue/Types.swift]: Yes, we should use `pkceBase64EncodedString`

Comment (grdsdev) [reply Sources/GoTrue/Types.swift]:
```suggestion
        return Data(buffer).pkceBase64EncodedString()
```

**Review by grdsdev — CHANGES_REQUESTED** [review]

**Review by grdsdev — APPROVED** [review]

**Comment by grdsdev** [comment]
Closing in favor of: https://github.com/supabase-community/supabase-swift/pull/119

**Pull request closed without merging**